src/consumer/offsetManager/manualOffsetManager.js

```diff
@@ -0,0 +1,145 @@
+  }
+
+  /**
+   * @param {string} topic
+   * @param {number} partition
+   * @returns {Long}
+   */
+  nextOffset(topic, partition) {
+    let offset = this.resolvedOffsets[topic][partition]
+    if (isInvalidOffset(offset)) {
+      offset = '0'
+    }
+
+    return Long.fromValue(offset)
+  }
+
+  /**
+   * @param {import("../../../types").TopicPartitionOffset} topicPartitionOffset
+   */
+  resolveOffset({ topic, partition, offset }) {
+    this.resolvedOffsets[topic][partition] = Long.fromValue(offset)
+      .add(1)
+      .toString()
+  }
+
+  /**
+   * @param {import("../../../types").TopicPartition} topicPartition
+   */
+  setDefaultOffset({ topic, partition }) {
+    const defaultOffset = this.cluster.defaultOffset(this.topicConfigurations[topic])
+    this.resolvedOffsets[topic][partition] = Long.fromValue(defaultOffset).toString()
+  }
+
+  /**
+   * Set the offset to be used for the next fetch. If the consumer isn't assigned to the given
+   * topic/partition this method will be a NO-OP.
+   *
+   * @param {import("../../../types").TopicPartitionOffset} topicPartitionOffset
+   */
+  async seek({ topic, partition, offset }) {
+    if (!this.memberAssignment[topic] || !this.memberAssignment[topic].includes(partition)) {
+      return
+    }
+
+    this.resolveOffset({
+      topic,
+      partition,
+      offset: Long.fromValue(offset)
+        .subtract(1)
+        .toString(),
+    })
+  }
+
+  /**
+   * resolveOffsets can be called when first initializing the consumer, in which case all partitions are pending,
+   * or when partitions have been added or removed. Newly added partitions are pending.
+   */
+  async resolveOffsets() {
+    const invalidOffset = topic => partition => {
+      return isInvalidOffset(this.resolvedOffsets[topic][partition])
+    }
+
+    const pendingPartitions = this.topics
+      .map(topic => ({
+        topic,
+        partitions: this.memberAssignment[topic]
+          .filter(invalidOffset(topic))
+          .map(partition => ({ partition })),
+      }))
+      .filter(t => t.partitions.length > 0)
+
+    if (pendingPartitions.length === 0) {
+      return
+    }
+
+    const unresolvedPartitions = pendingPartitions.map(({ topic, partitions }) =>
+      assign(
+        {
+          topic,
+          partitions,
+        },
+        this.topicConfigurations[topic]
+      )
+    )
+
+    const indexPartitions = (obj, { partition, offset }) => {
+      return assign(obj, { [partition]: offset })
+    }
+
+    const hasUnresolvedPartitions = () => unresolvedPartitions.some(t => t.partitions.length > 0)
+
+    if (hasUnresolvedPartitions()) {
+      const topicOffsets = await this.cluster.fetchTopicsOffset(unresolvedPartitions)
+      const offsets = initializeConsumerOffsets(topicOffsets, topicOffsets)
+      offsets.forEach(({ topic, partitions }) => {
+        this.resolvedOffsets[topic] = partitions.reduce(indexPartitions, {
+          ...this.resolvedOffsets[topic],
```

Comment:
can you simplify this ? a bit  hard to understand what going on here

Reply:
This merges the unresolved partitions into `this.resolvedOffsets`. I kept it like this for consistency with how we do the merging of partitions into `this.committedOffsets` in the existing group offset manager.